File: 295.findMedianFromDataStream/findMedianFromDataStream.java
class MedianFinder {

    /** initialize your data structure here. */
    public MedianFinder() {
        leftQ = new PriorityQueue<Integer>(1000, Collections.reverseOrder());
        rightQ = new PriorityQueue<Integer>();
    }
    
    public void addNum(int num) {
        leftQ.add(num);
        rightQ.add(leftQ.poll());
        if (leftQ.size() < rightQ.size()) {
            leftQ.add(rightQ.poll());
        }
        size++;
    }
    
    public double findMedian() {
        if (size % 2 == 0) {
            return (leftQ.peek() + rightQ.peek()) / 2.0;
        } else {
            return leftQ.peek();
        }
    }
    
    Queue<Integer> leftQ;
    Queue<Integer> rightQ;
    int size;
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: 295.findMedianFromDataStream/findMedianFromDataStream.cpp
// clang++ findMedianFromDataStream.cpp -std=c++11

#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class MedianFinder{
public:
  MedianFinder() {
    
  }
  void addNum(int num) {
    /*if (leftQ.empty() || num <= leftQ.top()) {
      leftQ.push(num);
      if (leftQ.size() - 2 == rightQ.size()) {
        rightQ.push(-leftQ.top());
        leftQ.pop();
      }
    } else {
      rightQ.push(-num);
      if (leftQ.size() < rightQ.size()) {
	leftQ.push(-rightQ.top());
	rightQ.pop();
	} 
    } */
    leftQ.push(num);
    rightQ.push(-leftQ.top());
    leftQ.pop();
    if (leftQ.size() < rightQ.size()) {
	leftQ.push(-rightQ.top());
	rightQ.pop();
    }
  }
  double findMedian() {
    return (leftQ.size() > rightQ.size()) ? leftQ.top() : (leftQ.top() - rightQ.top()) / 2.0;
  }
  priority_queue<long> leftQ, rightQ;
};

int main()
{
  MedianFinder mf;
  mf.addNum(1);
  mf.addNum(2);
  cout << mf.findMedian() << endl;
  mf.addNum(3);
  cout << mf.findMedian() << endl;
  return 0;
}
